File: src/vai_library/graph_runner/test/test_softmax.py
#
# Copyright 2022-2023 Advanced Micro Devices Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import unittest
import xir
import vart
import xir_extra_ops
import numpy as np
import simple_graph


class AllTest(unittest.TestCase):
    def test_plate_number(self):
        g = xir.Graph.deserialize("plate_num.xmodel")
        simple_graph.set_reference_input (g, "plate_number.golden/plate_data.bin")
        simple_graph.add_check_point (g, {("score1m", "download"):
                                          "plate_number.golden/score1m.bin"})
        simple_graph.normal_setting_for_graph (g)
        simple_graph.run_graph (g)
    def test_softmax(self):
        g = xir.Graph("test_softmax")
        data = np.asarray([[2.0, 2.0, 3.0, 4.0]], dtype=np.float32);
        data_op = g.create_const_op ("input", data)
        data_op.set_attr("subgraph_path", ["cpu1"])
        softmax_op = g.create_op("softmax",
                                 "softmax",
                                 attrs={"axis": -1,
                                        "subgraph_path": ["cpu1"]},
                                 input_ops= {"input": [data_op]});
        compare_op = g.create_op("compare",
                                 "compare",
                                 attrs={
                                     "from_file": "test/ref_results/softmax0.bin",
                                     "subgraph_path": ["cpu1"]
                                 },
                                 input_ops= {"input": [softmax_op]});
        simple_graph.create_simple_cpu_graph (g)
        simple_graph.run_graph (g)

if __name__ == '__main__':
    unittest.main()

File: src/vai_runtime/vart/dpu-controller/test/test_softmax.cpp
/*
 * Copyright 2022-2023 Advanced Micro Devices Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <cmath>
#include <fstream>
#include <future>
#include <iostream>
#include <thread>

#include "vitis/ai/parse_value.hpp"
#include "vitis/ai/profiling.hpp"
#include "xir/sfm_controller.hpp"

using namespace std;
DEF_ENV_PARAM(DEBUG_TEST, "0");

static void softmax_c(const int8_t* input, float scale, unsigned int cls,
                      float* output) {
  float sum = 0.f;
  for (unsigned int i = 0; i < cls; ++i) {
    output[i] = exp(input[i] * scale);
    sum += output[i];
  }
  for (unsigned int i = 0; i < cls; ++i) output[i] /= sum;
}

static void softmax_c(const int8_t* input, float scale, unsigned int cls,
                      unsigned int group, float* output) {
  for (unsigned int i = 0; i < group; ++i) {
    softmax_c(input, scale, cls, output);
    input += cls;
    output += cls;
  }
}

static void compare(int cls, int group, signed char* input, float* output1,
                    float* output2) {
  for (auto g = 0; g < group; ++g) {
    for (auto i = 0; i < cls; ++i) {
      auto idx = g * cls + i;
      auto diff = output1[idx] - output2[idx];
      if (ENV_PARAM(DEBUG_TEST) || (diff != 0.0 && std::abs(diff) > 0.001)) {
        cout << " i=" << i               //
             << " g = " << g             //
             << " idx = " << idx << " "  //
             << (int)input[idx] << ": " << output1[idx] << " " << output2[idx]
             << " " << diff << endl;
      }
    }
  }
}

int main(int argc, char* argv[]) {
  auto sfm = xir::SfmController::get_instance();
  if (argc < 5) {
    cout << "usage: " << argv[0] << "<file> <fixpos> <cls> <group>" << endl;
    return 0;
  }
  char* file = argv[1];
  int fixpos = stoi(argv[2]);
  int cls = stoi(argv[3]);
  int group = stoi(argv[4]);
  float scale = std::exp2f(-1.0f * (float)fixpos);
  vector<int8_t> input(cls * group);
  vector<float> output1(cls * group);
  vector<float> output2(cls * group);
  LOG_IF(INFO, ENV_PARAM(DEBUG_TEST))
      << " fixpos=" << fixpos << " cls=" << cls << " group=" << group
      << " scale=" << scale;
  CHECK(std::ifstream(file)
            .read(reinterpret_cast<char*>(&input[0]), input.size())
            .good())
      << "cannot read file " << file << " "
      << "cls " << cls << " "      //
      << "group " << group << " "  //
      << "scale " << scale << " "  //
      ;

  __TIC__(sfmx);
  sfm->run(&input[0], scale, cls, group, &output1[0]);
  __TOC__(sfmx);
  softmax_c(&input[0], scale, cls, group, &output2[0]);
  compare(cls, group, &input[0], &output1[0], &output2[0]);
  return 0;
}
